resources/assets/lib/stores/notification-stack-store.ts

```diff
@@ -0,0 +1,221 @@
+import DispatchListener from 'dispatch-listener';
+import NotificationJson, { NotificationBundleJson, NotificationStackJson, NotificationTypeJson } from 'interfaces/notification-json';
+import { action, computed, observable } from 'mobx';
+import LegacyPmNotification from 'models/legacy-pm-notification';
+import Notification from 'models/notification';
+import NotificationStack, { idFromJson } from 'models/notification-stack';
+import NotificationType, { Name as NotificationTypeName  } from 'models/notification-type';
+import { nameToCategory } from 'notification-maps/category';
+import { NotificationEventMoreLoaded, NotificationEventNew, NotificationEventRead } from 'notifications/notification-events';
+import { fromJson, NotificationIdentity, resolveIdentityType, resolveStackId } from 'notifications/notification-identity';
+import { NotificationResolver } from 'notifications/notification-resolver';
+import NotificationStore from './notification-store';
+
+@dispatchListener
+export default class NotificationStackStore implements DispatchListener {
+  @observable readonly legacyPm = new LegacyPmNotification();
+  @observable readonly types = new Map<string | null, NotificationType>();
+  private readonly resolver = new NotificationResolver();
+
+  get allStacks() {
+    return this.allType.stacks;
+  }
+
+  get allType() {
+    return this.getOrCreateType({ objectType: null });
+  }
+
+  get isEmpty() {
+    return this.types.size === 1 && this.allStacks.size === 0;
+  }
+
+  @computed get total() {
+    return this.allType.total;
+  }
+
+  set total(value: number) {
+    this.allType.total = value;
+  }
+
+  constructor(protected notificationStore: NotificationStore) {
+    // 'all' type should always exist; makes testing deterministic.
+    this.getOrCreateType({ objectType: null });
+  }
+
+  @action
+  flushStore() {
+    this.types.clear();
+  }
+
+  getOrCreateType(identity: NotificationIdentity) {
+    let type = this.types.get(identity.objectType);
+    if (type == null) {
+      type = new NotificationType(identity.objectType, this.resolver);
+      this.types.set(type.name, type);
+    }
+
+    return type;
+  }
+
+  getStack(identity: NotificationIdentity) {
+    return this.types.get(identity.objectType)?.stacks.get(resolveStackId(identity));
+  }
+
+  @action
+  handleDispatchAction(dispatched: DispatcherAction) {
+    if (dispatched instanceof NotificationEventNew) {
+      this.handleNotificationEventNew(dispatched);
+    } else if (dispatched instanceof NotificationEventMoreLoaded) {
+      this.handleNotificationEventMoreLoaded(dispatched);
+    } else if (dispatched instanceof NotificationEventRead) {
+      this.handleNotificationEventRead(dispatched);
+    } else if (dispatched instanceof UserLoginAction || dispatched instanceof UserLogoutAction) {
+      this.flushStore();
+    }
+  }
+
+  @action
+  handleNotificationEventMoreLoaded(event: NotificationEventMoreLoaded) {
+    if (!event.context.isWidget) {
```

Comment:
`event.context.store === this`

Reply:
params for notification event messages should be easily json serializable, so only json and simple data objects, or similar; it shouldn't be using things like stores, controllers, etc.